crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -393,10 +541,185 @@ impl Drop for MemoryPool {
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct StripedAllocationIndex(u32);
+
+impl StripedAllocationIndex {
+    fn from_unstriped_slot_index(index: MemoryAllocationIndex, num_stripes: usize) -> Self {
+        let num_stripes: u32 = num_stripes.try_into().unwrap();
+        Self(index.0 / num_stripes)
+    }
+
+    fn as_unstriped_slot_index(self, stripe: usize, num_stripes: usize) -> MemoryAllocationIndex {
+        let num_stripes: u32 = num_stripes.try_into().unwrap();
+        let stripe: u32 = stripe.try_into().unwrap();
+        MemoryAllocationIndex(self.0 * num_stripes + stripe)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct SlabConstraints {
+    max_memory_bytes: usize,
+    num_slots: usize,
+    expected_slot_bytes: usize,
+    num_pkeys_available: usize,
+    guard_bytes: usize,
+    guard_before_slots: bool,
+}
+
+#[derive(Debug)]
+struct SlabLayout {
+    /// The total number of slots available in the memory pool slab.
+    num_slots: usize,
+    /// The size of each slot in the memory pool; this contains the maximum
+    /// memory size (i.e., from WebAssembly or Wasmtime configuration) plus any
+    /// guard region after the memory to catch OOB access. On these guard
+    /// regions, note that:
+    /// - users can configure how aggressively (or not) to elide bounds checks
+    ///   via `Config::static_memory_guard_size`
+    /// - memory protection keys can compress the size of the guard region by
+    ///   placing slots from a different key (i.e., a stripe) in the guard
+    ///   region
+    slot_bytes: usize,
+    // The maximum size that can become accessible, in bytes, for each linear
+    // memory. Guaranteed to be a whole number of wasm pages.
+    max_memory_bytes: usize,
+    /// The total number of bytes to allocate for the memory pool slab.
+    // total_slab_bytes: usize,
+    /// If necessary, the number of bytes to reserve as a guard region at the
+    /// beginning of the slab.
+    pre_slab_guard_bytes: usize,
+    /// Like `pre_slab_guard_bytes`, but at the end of the slab.
+    post_slab_guard_bytes: usize,
+    /// The number of stripes needed in the slab layout.
+    num_stripes: usize,
+}
+
+impl SlabLayout {
+    /// Return the total size of the slab, using the final layout (where `n =
+    /// num_slots`):
+    ///
+    /// ```text
+    /// ┌────────────────────┬──────┬──────┬───┬──────┬─────────────────────┐
+    /// │pre_slab_guard_bytes│slot 1│slot 2│...│slot n│post_slab_guard_bytes│
+    /// └────────────────────┴──────┴──────┴───┴──────┴─────────────────────┘
+    /// ```
+    fn total_slab_bytes(&self) -> Result<usize> {
+        self.slot_bytes
+            .checked_mul(self.num_slots)
+            .and_then(|c| c.checked_add(self.pre_slab_guard_bytes))
+            .and_then(|c| c.checked_add(self.post_slab_guard_bytes))
+            .ok_or_else(|| anyhow!("total size of memory reservation exceeds addressable memory"))
```

Comment:
Also reading further I see now that this is infallible if `calculate` returns `Ok`.

In that case I think it's fine to leave this as-is, I see how it's hard to deduplicate otherwise.